test/computDYsplitting_PT.cpp

```diff
@@ -0,0 +1,211 @@
+
+  // input and output setup
+  // ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ----
+  TChain * bigChain = new TChain ("HTauTauTree/HTauTauTree") ;
+  appendFromFileList (bigChain, inputFile);
+  bigTree theBigTree (bigChain) ;
+
+  int totEvents = 0;
+  int allEvts[5][5][6]; // njet,nb,ht
+  for (int i = 0; i < 5; i++)
+    for (int j = 0; j < 5; j++)
+      for (int k = 0; k < 6; k++)
+	allEvts[i][j][k] = 0;
+
+  float xbins[6]  = {0.0, 1.0, 2.0, 3.0, 4.0, 5.0};
+  float ybins[6]  = {0.0, 1.0, 2.0, 3.0, 4.0, 5.0};
+  float zbins[7]  = {0.0, 50., 100., 250., 400., 650., 10000.};
+
+  TFile* fOut = new TFile (outputFileName.c_str(), "recreate");
+  TH3F* h_nJets_nBs_pt    = new TH3F ("h_nJets_nBs_pt"   , ";nJets;nBs;pt", 5, xbins, 5, ybins, 6, zbins);
+  TH3I* hINT_nJets_nBs_pt = new TH3I ("hINT_nJets_nBs_pt", ";nJets;nBs;pt", 5, xbins, 5, ybins, 6, zbins);
+
+  // speed up
+  theBigTree.fChain->SetBranchStatus("*", 0);
+  theBigTree.fChain->SetBranchStatus("lheNOutPartons", 1);
+  theBigTree.fChain->SetBranchStatus("lheNOutB", 1);
+  //theBigTree.fChain->SetBranchStatus("lheHt", 1);
+  theBigTree.fChain->SetBranchStatus("genpart_px",1);
+  theBigTree.fChain->SetBranchStatus("genpart_py",1);
+  theBigTree.fChain->SetBranchStatus("genpart_pz",1);
+  theBigTree.fChain->SetBranchStatus("genpart_e",1);
+  theBigTree.fChain->SetBranchStatus("genpart_flags",1);
+  theBigTree.fChain->SetBranchStatus("genpart_pdg",1);
+
+
+  cout << "** INFO: most branches are desactivated to speed up reading..." << endl;
+  // loop over events
+  // ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ---- ----
+  for (Long64_t iEvent = 0 ; true ; ++iEvent)
+  {
+    if (iEvent % 10000 == 0)  cout << "- reading event " << iEvent << endl ;
+
+    if (DEBUG) cout << "====== new event ====== " << iEvent << endl ;
+
+    int got = theBigTree.fChain->GetEntry(iEvent);
+    if (got == 0) break;
+
+    ++totEvents ;
+
+    int npartons = theBigTree.lheNOutPartons ;
+    if (npartons > 4)
+    {
+      cout << " --> ev : " << iEvent << " npartons is " << npartons << " lowered to 4" << endl;
+      npartons = 4;
+    }
+
+    if (DEBUG) cout << "n partons: " << theBigTree.lheNOutPartons << endl;
+
+    int nbs = theBigTree.lheNOutB;
+    if (nbs > 4)
+    {
+      cout << " --> ev : " << iEvent << " nbs is " << nbs << " lowered to 4" << endl;
+      nbs = 4;
+    }
+
+    if (DEBUG) cout << "NBs = " << nbs << endl;
+
+    //float ht = theBigTree.lheHt;
+    //if (ht >= 10000)
+    //{
+    //  cout << " --> ev: " << iEvent << " ht is " << ht << " lowered to 9999" << endl;
+    //  ht = 9999.;
+    //}
+    
+    // loop through gen parts to identify Z boson
+    int idx = -1;
+    for (unsigned int igen = 0; igen < theBigTree.genpart_px->size(); igen++)
+    {
+      bool isLast   = CheckBit(theBigTree.genpart_flags->at(igen), 13) ; // 13 = isLastCopy
+      bool isPrompt = CheckBit(theBigTree.genpart_flags->at(igen),  0) ; //  0 = isPrompt
```

Comment:
IIUC we want to replace this line with ```bool isFromHardProcess = CheckBit(theBigTree.genpart_flags->at(igen),  8) ; //  8 = isFromHardProcess```

Reply:
rather adding the line you're suggesting below this I would say (?)

Reply:
Yes, exactly, checking isLast and isFromHardProcess, at least this what @kramerto was proposing.

Reply:
Apologies, after further checking I realize I made a mistake and we should use * isPrompt && isFromHardProcess*

Reply:
My proposal was to use isPrompt && fromHardProcess **OS lepton pairs**, though, because we saw cases where there is no pdg 23 particle (so maybe only for those cases), if I recall correctly.